File: 06ros/ws02/third_ws/src/use_cus_action_01_pkg/scripts/client_simple_action.py
#! /usr/bin/env python
# coding=utf8

import rospy
from actionlib import SimpleActionClient, GoalStatus
from custom_action_msgs.msg import (
    CountNumberAction, CountNumberGoal, CountNumberFeedback,
    CountNumberResult
)


def done_cb(state, result):
    # if not isinstance(state,GoalStatus):
    #    return

    # if not isinstance(result,CountNumberResult):
    #    return

    if state == GoalStatus.ABORTED:
        rospy.loginfo("server working error, don't finish my job.")
    elif state == GoalStatus.PREEMPTED:
        rospy.loginfo("client cancel job.")
    elif state == GoalStatus.SUCCEEDED:
        rospy.loginfo("server finish job.")
        rospy.loginfo("[done] result:%f" % result.result)


def active_cb():
    rospy.loginfo("[active]")


def feedback_cb(feedback):
    # if not isinstance((feedback,CountNumberFeedback)):
    #    return

    rospy.loginfo("[feedback] percent:%f %%" % (feedback.rate * 100))


if __name__ == '__main__':
    nodeName = "client_node"
    rospy.init_node(nodeName)

    actionName = "/zcb01/action"
    client = SimpleActionClient(actionName, CountNumberAction)
    client.wait_for_server()

    goal = CountNumberGoal()
    goal.max = 101
    goal.duration = 0.1
    client.send_goal(goal, done_cb, active_cb, feedback_cb)

    # d = rospy.Duration(3)
    # rospy.sleep(d)
    # client.cancel_goal()

    rospy.spin()

File: 06ros/ws02/third_ws/src/use_cus_action_01_pkg/src/client_simple_action.cpp
//
// Created by zcb on 5/4/21.
//
#include <actionlib/client/simple_action_client.h>
#include <custom_action_msgs/CountNumberAction.h>
#include <ros/ros.h>

void done_callback(
        const actionlib::SimpleClientGoalState &state,
        const custom_action_msgs::CountNumberResultConstPtr &result
) {
    if (state == state.ABORTED) {
        ROS_INFO("server working error, don't finish my job.");
    } else if (state == state.PREEMPTED) {
        ROS_INFO("client cancel job.");
    } else if (state == state.SUCCEEDED) {
        ROS_INFO("server finish job.");
        ROS_INFO_STREAM("[done] result:" << result->result);
    }


}

void active_callback() {
    ROS_INFO_STREAM("[active]");
}

void fb_callback(
        const custom_action_msgs::CountNumberFeedbackConstPtr &feedback) {
    ROS_INFO_STREAM("[feedback] percent:" << feedback->rate * 100 << "%");
}


int main(int argc, char **argv) {
    std::string nodeName = "client_node";
    ros::init(argc, argv, nodeName);

    ros::NodeHandle node;

    std::string actionName = "/zcb01/action";
    actionlib::SimpleActionClient<custom_action_msgs::CountNumberAction> client(node, actionName);
    client.waitForServer();

    custom_action_msgs::CountNumberGoal goal;
    goal.max = 111;
    goal.duration = 0.1;

    client.sendGoal(goal, done_callback, active_callback, fb_callback);

//    ros::Duration(3).sleep(); // 3 s 后 取消任务
//    client.cancelGoal();

    ros::spin();

    return 0;
}
